overlord/devicestate/devicemgr.go

```diff
@@ -1425,6 +1425,73 @@ func (m *DeviceManager) ensurePostFactoryReset() error {
 	return os.Remove(factoryResetMarker)
 }
 
+func (m *DeviceManager) hasUserExpired(user *auth.UserState) bool {
+	// If the user has no expiration date, then Expiration should not
+	// be set, and contain the default value.
+	if user.Expiration.IsZero() {
+		return false
+	}
+
+	return user.Expiration.Before(time.Now())
+}
+
+var osutilDelUser = osutil.DelUser
+
+func (m *DeviceManager) removeExpiredUser(st *state.State, user *auth.UserState) error {
+	// Remove the user from the system first
+	if err := osutilDelUser(user.Username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
+		return err
+	}
+
+	// And then we remove it from the auth state
+	_, err := auth.RemoveUserByUsername(st, user.Username)
+	return err
+}
+
+// ensureExpiredUsersRemoved is periodically called as a part of Ensure()
+// to remove expired users from the system.
+func (m *DeviceManager) ensureExpiredUsersRemoved() error {
+	st := m.state
+	st.Lock()
+	defer st.Unlock()
+
+	// XXX: Only on ubuntu core systems?
```

Comment:
As far as I can tell from other parts of the code, it seems to me that "on classic" only changes the location of the users' database (on core, we need to use the `--extrausers` flag), but the fact that users are added remains valid on classic too. But better wait for Samuele.

Reply:
I agree! @pedronis

Reply:
I removed the check as you said the users added remain valid on classic, I instead refactored the test to verify that `--extrausers` flag is correctly set/not set on core/classic.